truffle/src/main/java/org/jruby/truffle/stdlib/readline/ReadlineHistoryNodes.java

```diff
@@ -0,0 +1,279 @@
+
+            return taintNode.executeTaint(ret);
+        }
+
+    }
+
+    @CoreMethod(names = { "length", "size" }, needsSelf = false)
+    public abstract static class LengthNode extends CoreMethodArrayArgumentsNode {
+
+        @TruffleBoundary
+        @Specialization
+        public int length() {
+            return getContext().getConsoleHolder().getHistory().size();
+        }
+
+    }
+
+    @CoreMethod(names = "clear", needsSelf = false)
+    public abstract static class ClearNode extends CoreMethodArrayArgumentsNode {
+
+        @TruffleBoundary
+        @Specialization
+        public DynamicObject clear() {
+            getContext().getConsoleHolder().getHistory().clear();
+
+            return nil();
+        }
+
+    }
+
+    @CoreMethod(names = "each", needsBlock = true)
+    public abstract static class EachNode extends YieldingCoreMethodNode {
+
+        @Child private TaintNode taintNode;
+
+        public EachNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            taintNode = TaintNodeGen.create(context, sourceSection, null);
+        }
+
+        @Specialization
+        public DynamicObject each(VirtualFrame frame, DynamicObject history, DynamicObject block) {
+            final ConsoleHolder consoleHolder = getContext().getConsoleHolder();
+
+            for (Iterator<History.Entry> i = consoleHolder.getHistory().iterator(); i.hasNext();) {
+                final DynamicObject line = createString(StringOperations.createRope(i.next().value().toString(), getDefaultInternalEncoding()));
+
+                yield(frame, block, taintNode.executeTaint(line));
+            }
+
+            return history;
+        }
+
+    }
+
+    @CoreMethod(names = "[]", needsSelf = false, required = 1, lowerFixnumParameters = 0)
+    public abstract static class GetIndexNode extends CoreMethodArrayArgumentsNode {
+
+        @Child private TaintNode taintNode;
+
+        public GetIndexNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            taintNode = TaintNodeGen.create(context, sourceSection, null);
+        }
+
+        @TruffleBoundary
+        @Specialization
+        public Object getIndex(int index) {
+            final ConsoleHolder consoleHolder = getContext().getConsoleHolder();
+
+            final int normalizedIndex = index < 0 ? index + consoleHolder.getHistory().size() : index;
+
+            try {
+                final String line = consoleHolder.getHistory().get(normalizedIndex).toString();
+                final DynamicObject ret = createString(StringOperations.createRope(line, getDefaultInternalEncoding()));
+
+                return taintNode.executeTaint(ret);
+            } catch (IndexOutOfBoundsException e) {
+                throw new RaiseException(coreExceptions().indexError("invalid index", this));
```

Comment:
Could you make an `indexErrorInvalidIndex` method? I'm trying to keep all the string constants in `RubyException` (and they'll go into `CoreStrings` eventually).